src/harness/reference_models/iap/iap.py

```diff
@@ -0,0 +1,623 @@
+    and aggregate_interference instance of class type AggregateInterferenceOutputFormat.
+    It is updated in the format of {latitude : { longitude : [a_p(mW/IAPBW),..] }}
+  """
+  
+  # Get all the grants inside neighborhood of the protection entity
+  grants_inside = interf.findGrantsInsideNeighborhood(
+                    grant_objects, protection_point, protection_ent_type)
+   
+  if len(grants_inside) > 0:
+    # Get protection constraint of protected entity
+    protection_constraint = interf.ProtectionConstraint(latitude=protection_point[1],
+                               longitude=protection_point[0], low_frequency=low_freq,
+                               high_frequency=high_freq, entity_type=protection_ent_type)
+
+    # Get all the grants fully/partially overlap with protection point,p
+    neighborhood_grants = interf.findOverlappingGrants(
+                            grants_inside, protection_constraint)
+
+    num_unsatisfied_grants = len(neighborhood_grants)
+    
+    # Get number of grants, IAP protection threshold and fairshare per channel 
+    if num_unsatisfied_grants > 0:
+      # algorithm to calculate interference within IAPBW using EIRP obtained for all 
+      # the grants through application of IAP
+
+      # Initialize list of aggregate interference from all the grants 
+      # ( including grants from managing and peer SAS )
+      aggr_interf = [0] * len(channels)
+      # Initialize list of aggregate interference from all the grants 
+      # ( including only grants from managing SAS )
+      asas_interf = [0] * len(channels)
+
+      num_unsatisfied_grants_channel = []
+      iap_threshold_channel = []
+      fairshare_channel = []
+
+      for index, channel in enumerate(channels):
+        # Get protection constraint over 5MHz channel range
+        channel_constraint = interf.ProtectionConstraint(latitude=protection_point[1],
+                               longitude=protection_point[0], low_frequency=channel[0],
+                               high_frequency=channel[1], entity_type=protection_ent_type)
+
+        # Identify CBSD grants overlapping with frequency range of the protection entity 
+        # in the neighborhood of the protection point and channel
+        channel_neighborhood_grants = interf.findOverlappingGrants(
+                                        neighborhood_grants, channel_constraint)
+
+        num_unsatisfied_grants_channel.append(len(channel_neighborhood_grants))
+        
+        # Set interference quota for protection_point, channel combination
+        # Calculate Roll of Attenuation for ESC Sensor Protection entity
+        if protection_ent_type is interf.ProtectedEntityType.ESC:
+          if channel[0] >= 3650.e6:
+            center_freq = (channel[0] + channel[1]) / 2
+            iap_threshold_channel.append(interf.dbToLinear(interf.linearToDb(threshold) + 
+              (2.5 + ((center_freq - interf.ESC_CH21_CF_HZ) / MHZ))))
```

Comment:
Agreed with Serge that you need to adjust interference (not threshold). They are not equivalent I think.

In addition, is this the logic to handle ESC Rx Filter Loss? If so, I don't quite understand this piece of code. How did you come up with this formula? Can you explain? I'm afraid that it is not consistent with the spec.

Reply:
2.5 dB assumes the filter loss is determined by the center frequency. To be more accurate, we should integrate over the 5 MHz frequency range. If we do the integration by averaging the filter loss over five 1 MHz frequency range, the filter loss should be 2.27 dB. That means using 2.5 dB slightly under-protects ESC. The rest of the equation can say the same.

Reply:
R2-SGN-25   SAS ESC Sensor Protection     a. SASs shall manage CBRS interference for all ESC sensors that require   protection such that the aggregate mean interference at the reference ESC   filter output of the protected sensor in 3550-3700 MHz does not exceed - 109   dBm/MHz. The reference ESC filter has 0.5 dB insertion loss in the   passband     b. For initial certification, the SAS shall treat Category B CBSDs   operating within the frequency range 3650-3680 MHz, and Category A CBSDs   operating within the frequency range 3650-3660 MHz, as co-channel to ESCs and   apply the same protection described above, after assuming a straight line 1   dB per MHz ESC reference filter roll-off from 3650-3680 MHz          Our interpretation of the requirement is that the IAP calculations should   be applied without modification after adjusting for the ESC reference filter   roll-off. This is accomplished by adjusting the threshold (Q) to compensate   for the roll-off, i.e., increasing it by an amount corresponding to the   roll-off. Then the IAP algorithm (same protection described above) can be   applied. Modifying the individual grant interference contributions based on   the rolloff would be a modification of the "protection described   above." Q represents the acceptable amount of aggregate interference at   the ESC filter input. The filter then attenuates the aggregate by the   rolloff.

Reply:
For adjusting the threshold for the roll-off, if we are calculating the average aggregate over 1 MHz and then using that to calculate the over-all attenuation over the 5MHz BW, we need to integrate the average attenuation value over the BW, which is not being done here. Also to compute the average value, we have to find the mean value of (1/attn_per_mhz) over the 1 MHz points in the IAP Reference BW where attn_per_MHz > 0. So if we take an example of the channel 3650 to 3655, the average attenuation would be around -1.7723 dB.

Reply:
Agreed, updated the code with   the integration over the five 1MHZ frequency logic to calculate the ESC   rolloff